[ogre-next] Fixup 2.3.3

dg0yt opened the pull request:
Cleanup installation. This should have been part of 2.3.3#0.
Don't pull in unused threading backend dependencies. (Upstream CI runs with threading backend disabled. If you want threading, add a feature which supplies the necessary dependencies and options.)
Fix platform support globally and for rendering backends which are now modeled as features.
The vulkan backend builds locally in x64-linux. vcpkg CI lacks the SDK. This also blocks CI for android.

@BillyONeal This was nearly ready when you asked https://github.com/microsoft/vcpkg/pull/34945#issuecomment-1839821202. It was in draft state at that time because I had to wait for related PRs to be merged to have CI as needed for this port.

dg0yt committed:
[ogre-next] Rebase changes

dg0yt committed:
Fixup

dg0yt committed:
fixup

talregev commented:
Why you remove the optional dependencies? 
Can you add them as a features?

dg0yt commented:
> Why you removed the optional dependencies? Can you add them as a features?

This PR removes unused dependencies and adds necessary dependencies.
I have no need to add any other features.

JonLiu1993 commented:
@dg0yt, Thanks for your PR, when I tested the feature by command `./vcpkg i./vcpkg install ogre-next[d3d11,gl3plus,planar-reflections,vulkan]nstall ogre-next[d3d11,gl3plus,planar-reflections,vulkan]:x86-windows`, I get this error:
````
-- Building x64-windows-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: F:/Feature-test/orge/downloads/tools/cmake-3.27.1-windows/cmake-3.27.1-windows-i386/bin/cmake.exe --build . --config Debug --target install -- -v -j9
    Working Directory: F:/Feature-test/orge/buildtrees/ogre-next/x64-windows-dbg
    See logs for more information:
      F:\Feature-test\orge\buildtrees\ogre-next\install-x64-windows-dbg-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  ports/ogre-next/portfile.cmake:91 (vcpkg_cmake_install)
  scripts/ports.cmake:170 (include)
````
F:\Feature-test\orge\buildtrees\ogre-next\install-x64-windows-dbg-out.log:
````
FAILED: bin/RenderSystem_Vulkan_d.dll RenderSystems/Vulkan/lib/RenderSystem_Vulkan_d.lib 
cmd.exe /C "cd . && F:\Feature-test\orge\downloads\tools\cmake-3.27.1-windows\cmake-3.27.1-windows-i386\bin\cmake.exe -E vs_link_dll --intdir=RenderSystems\Vulkan\CMakeFiles\RenderSystem_Vulkan.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1437~1.328\bin\Hostx64\x64\link.exe  @CMakeFiles\RenderSystem_Vulkan.rsp  /out:bin\RenderSystem_Vulkan_d.dll /implib:RenderSystems\Vulkan\lib\RenderSystem_Vulkan_d.lib /pdb:bin\RenderSystem_Vulkan_d.pdb /dll /version:2.3 /machine:x64 /nologo    /debug /INCREMENTAL  && cd ."
LINK Pass 1: command "C:\PROGRA~1\MIB055~1\2022\COMMUN~1\VC\Tools\MSVC\1437~1.328\bin\Hostx64\x64\link.exe @CMakeFiles\RenderSystem_Vulkan.rsp /out:bin\RenderSystem_Vulkan_d.dll /implib:RenderSystems\Vulkan\lib\RenderSystem_Vulkan_d.lib /pdb:bin\RenderSystem_Vulkan_d.pdb /dll /version:2.3 /machine:x64 /nologo /debug /INCREMENTAL /MANIFEST /MANIFESTFILE:RenderSystems\Vulkan\CMakeFiles\RenderSystem_Vulkan.dir/intermediate.manifest RenderSystems\Vulkan\CMakeFiles\RenderSystem_Vulkan.dir/manifest.res" failed (exit code 1120) with the following output:
   Creating library RenderSystems\Vulkan\lib\RenderSystem_Vulkan_d.lib and object RenderSystems\Vulkan\lib\RenderSystem_Vulkan_d.exp
OgreVulkanProgram.cpp.obj : error LNK2019: unresolved external symbol "public: __cdecl glslang::TShader::TShader(enum EShLanguage)" (??0TShader@glslang@@QEAA@W4EShLanguage@@@Z) referenced in function "public: bool __cdecl Ogre::VulkanProgram::compile(bool,bool)" (?compile@VulkanProgram@Ogre@@QEAA_N_N0@Z)
OgreVulkanProgram.cpp.obj : error LNK2019: unresolved external symbol "public: virtual __cdecl glslang::TShader::~TShader(void)" (??1TShader@glslang@@UEAA@XZ) referenced in function "public: bool __cdecl Ogre::VulkanProgram::compile(bool,bool)" (?compile@VulkanProgram@Ogre@@QEAA_N_N0@Z)
OgreVulkanProgram.cpp.obj : error LNK2019: unresolved external symbol "public: void __cdecl glslang::TShader::setStrings(char const * const *,int)" (?setStrings@TShader@glslang@@QEAAXPEBQEBDH@Z) referenced in function "public: bool __cdecl Ogre::VulkanProgram::compile(bool,bool)" (?compile@VulkanProgram@Ogre@@QEAA_N_N0@Z)
OgreVulkanProgram.cpp.obj : error LNK2019: unresolved external symbol "public: void __cdecl glslang::TShader::setEntryPoint(char const *)" (?setEntryPoint@TShader@glslang@@QEAAXPEBD@Z) referenced in function "public: bool __cdecl Ogre::VulkanProgram::compile(bool,bool)" (?compile@VulkanProgram@Ogre@@QEAA_N_N0@Z)
OgreVulkanProgram.cpp.obj : error LNK2019: unresolved external symbol "public: bool __cdecl glslang::TShader::parse(struct TBuiltInResourc
```
Please take a look.

dg0yt committed:
[glslang] Burn-in DLL linkage

dg0yt committed:
[glslang] Update path replacement

dg0yt committed:
[glslang] Drop broken legacy files

These files are not used by the CMake package, and they
wanted to include the target files from the lib/cmake/fixup.
If nobody noticed the problem, nobody will miss them.

dg0yt committed:
versions

dg0yt commented:
@JonLiu1993 Please test the update. Fixed the normal dllimport quirk, this time for port glslang.

dg0yt commented on ports/ogre-next/vcpkg.json:
```diff
@@ -30,9 +25,46 @@
+      "description": "Vulkan render system",
+      "dependencies": [
+        "glslang",
+        "vulkan"
```
Note https://github.com/microsoft/vcpkg/pull/35479#issuecomment-1858765149.

JonLiu1993 commented:
`./vcpkg install ogre-next[d3d11,gl3plus,planar-reflections,vulkan]:x86-windows`
````
F:\Feature-test\orge\buildtrees\ogre-next\src\v2.3.3-d8d38bba95.clean\RenderSystems\Vulkan\src\OgreVulkanRenderSystem.cpp(3363): error C2440: '=': cannot convert from 'VkSampler' to 'void *'
F:\Feature-test\orge\buildtrees\ogre-next\src\v2.3.3-d8d38bba95.clean\RenderSystems\Vulkan\src\OgreVulkanRenderSystem.cpp(3363): note: Conversion from integral type to pointer type requires reinterpret_cast, C-style cast or parenthesized function-style cast
F:\Feature-test\orge\buildtrees\ogre-next\src\v2.3.3-d8d38bba95.clean\RenderSystems\Vulkan\src\OgreVulkanRenderSystem.cpp(3369): error C2440: 'static_cast': cannot convert from 'void *' to 'VkSampler'
F:\Feature-test\orge\buildtrees\ogre-next\src\v2.3.3-d8d38bba95.clean\RenderSystems\Vulkan\src\OgreVulkanRenderSystem.cpp(3369): note: There is no context in which this conversion is possible
````
./vcpkg install ogre-next[vulkan]:arm64-android:
````
Stored binaries in 1 destinations in 69.7 ms.
Elapsed time to handle tinyxml:arm64-android: 2 s
Installing 26/28 vulkan:arm64-android...
Building vulkan:arm64-android...
-- Configuring arm64-android
CMake Error at ports/vulkan/portfile.cmake:30 (message):
  The Vulkan SDK wasn't found.
Call Stack (most recent call first):
  scripts/ports.cmake:170 (include)

````

dg0yt commented:
> CMake Error at ports/vulkan/portfile.cmake:30 (message):
>   The Vulkan SDK wasn't found.

Expected in default API level. (vulkan is a default feature for android, and that is the reaon for the ci.baseline entries.) 
Cf. #35725 for a CI run with API level hacks. 
With current port `vulkan`, test locally with
~~~
vcpkg install ogre-next[vulkan]:arm64-android --cmake-args="-DVCPKG_CMAKE_SYSTEM_VERSION=26"
~~~
(Almost literal mention in https://github.com/microsoft/vcpkg/pull/35479#issuecomment-1858374863.)

JonLiu1993 commented:
Use command `vcpkg install ogre-next[vulkan]:arm64-android --cmake-args="-DVCPKG_CMAKE_SYSTEM_VERSION=26"` can successfully test the `vulkan` feature.

BillyONeal commented:
Thanks~
